src/lib/io/sol-memmap-storage.c

```diff
@@ -16,58 +16,37 @@
  * limitations under the License.
  */
 
-#include "sol-memmap-storage.h"
+#include "sol-util-internal.h"
+
+#include "sol-memmap-storage-impl.h"
 
-#include <errno.h>
-#include <fcntl.h>
-#include <stdio.h>
-#include <sys/stat.h>
-#include <unistd.h>
+struct sol_log_domain _sol_memmap_storage_log_domain;
 
 #include "sol-buffer.h"
-#include "sol-log.h"
 #include "sol-mainloop.h"
-#include "sol-str-slice.h"
-#include "sol-str-table.h"
-#include "sol-util-file.h"
 #include "sol-vector.h"
-#include "sol-util-internal.h"
-
-#ifdef USE_I2C
-#include <sol-i2c.h>
-#endif
 
-#define REL_PATH_IDX 2
-#define DEV_NUMBER_IDX 3
-#define DEV_NAME_IDX 4
-
-struct pending_write_data {
-    char *name;
-    struct sol_blob *blob;
-    const struct sol_memmap_entry *entry;
-    void (*cb)(void *data, const char *name, struct sol_blob *blob, int status);
-    const void *data;
-    uint64_t mask;
-};
-
-struct map_internal {
-    const struct sol_memmap_map *map;
-    struct sol_timeout *timeout;
-    char *resolved_path;
-    struct sol_vector pending_writes;
-    bool checked;
-};
-
-static struct sol_vector memory_maps = SOL_VECTOR_INIT(struct map_internal);
+static struct sol_ptr_vector memory_maps = SOL_PTR_VECTOR_INIT;
+static bool initialised = false;
 
 static bool
-get_entry_metadata_on_map(const char *name, const struct sol_memmap_map *map, const struct sol_memmap_entry **entry, uint64_t *mask)
+get_entry_metadata_on_map(const char *name,
+    const char **out_name,
+    const struct sol_memmap_map *map,
+    const struct sol_memmap_entry **entry,
+    uint64_t *mask)
 {
     uint32_t bit_size;
+    const struct sol_str_table_ptr *table_entry = sol_str_table_ptr_entry_lookup
```

Comment:
why did you introduce this function if all you're looking for is the `table_entry->val`?

note that you can use the `errno` to detect if `NULL` is returned due problems or an empty pointer in the value.